### ISSUE TITLE
WebAudio on iOS

### ISSUE THREAD

**Issue by ffdd** [emscripten-ports/SDL2#19]
I'm working on trying to get some audio out of an SDL2 program and it works fine in most browsers except for Mobile Safari and Chrome on iOS - I believe it has something to do with: https://developer.apple.com/library/safari/documentation/AudioVideo/Conceptual/Using_HTML5_Audio_Video/PlayingandSynthesizingSounds/PlayingandSynthesizingSounds.html#//apple_ref/doc/uid/TP40009523-CH6-SW6

- I've tried initializing the SDL2 audio objects on a touch event but things are still silent. Any ideas?

**Comment by kripken** [comment]
Do web audio tests work in general on iOS?

**Comment by ffdd** [comment]
Web Audio in general works. I'm a bit new to SDL in general - are there some specific tests in the repo I could compile and try running on iOS then get back to you?

**Comment by ffdd** [comment]
It looks like the webkitAudioContext is 'locked' by default on iOS - https://paulbakaus.com/tutorials/html5/web-audio-on-ios/#How_do_I_know_when_my_context_is_unlocked -

**Comment by kripken** [comment]
Not sure what "locked" means. Is there something we can do?

**Comment by ffdd** [comment]
Basically on iOS in Chrome and Safari you can only play audio after a user interaction like a button press. I've gotten around this when developing websites by checking for the first touchdown or click in JavaScript and starting up the audio at that point if it wasn't already playing. This way you don't have to check the user-agent.

**Comment by Daft-Freak** [comment]
So, it looks like in order to "unlock" the audio context we have to call `noteOn` inside an event handler. I guess we'll have to create a dummy source to handle that as we're currently using `ScriptProcessorNode`

**Comment by ffdd** [comment]
Yeah I think you are right. I set up a basic example presuming that I could
work around this by booting up all the audio stuff on an SDL_EVENT but it
doesn't work on iOS browsers.

Maybe this file can be modified and used as a test since it will play audio
on click while compiling with `gcc` for desktop or with desktop browsers
(Chrome, Safari, Firefox) via `emscripten`. I'm a bit new to C in general
but I'm really familiar with JavaScript so let me know if there is anything
I can do to help get this done.

https://gist.github.com/a9d4f78345de0f79ed7e

rey.sc

**Comment by ffdd** [comment]
Ignore that gist - here is a better one:
https://gist.github.com/ffdd/1c9083fefc4f9188d5c4

rey.sc

**Comment by Daft-Freak** [comment]
Gist with a workaround:
https://gist.github.com/Daft-Freak/f22bf2e5447d8c56747e

The important bit is here: https://gist.github.com/Daft-Freak/f22bf2e5447d8c56747e#file-ios-audio-c-L52

**Comment by ffdd** [comment]
Ah it works beautifully! Thanks so much Charlie. I'm not exactly sure why
the SDL_EventFilter is necessary to make it work, even after reading the
docs, but thanks regardless.

rey.sc

**Comment by Daft-Freak** [comment]
The event filter is called while still inside the event handler and the event queue isn't. I'll see if I can get SDL to handle this automatically.

**Comment by ffdd** [comment]
The future of this on iOS probably isn't too predictable... I'm wondering
things like whether iOS will eventually support the HTML5 Gamepad API and
pressing a button or any key on the keyboard will unlock audio also, or if
browser games become more popular then Safari may send a request to the
user in the same way that browser's ask to use your location in order to
play audio automatically. Obviously these ideas aren't in the HTML5 spec
but neither was locking WebAudio before user interaction so who knows what
they will do. I could see SDL handling this magically, or just as easily
see this as a snippet in a wiki page for the repository about getting audio
to work on mobile browsers.

rey.sc

**Comment by ffdd** [comment]
I ended up using this code https://gist.github.com/ffdd/44075ddf97a5219dd3e5 in my project, where I wrap `include "mobile-audio.h"` and `mobile_audio_start_on_user_input();` in preprocessor directives for emscripten.
